[css-ui-4] Transparent accent colors (275767)

lilyspiniolas opened the pull request:
#### 65d8059e3250b90b4ed2cce6303d4df55edc2627
<pre>
[css-ui-4] Transparent accent colors (275767)
<a href="https://bugs.webkit.org/show_bug.cgi?id=275767">https://bugs.webkit.org/show_bug.cgi?id=275767</a>
<a href="https://rdar.apple.com/130599744">rdar://130599744</a>

Reviewed by Tim Nguyen and Abrar Rahman Protyasha.

Transparent accent-colors are now precomposed over the canvas to conform to <a href="https://github.com/w3c/csswg-drafts/issues/9852#issuecomment-2166096835.">https://github.com/w3c/csswg-drafts/issues/9852#issuecomment-2166096835.</a> Added transparent-accent-color wpt tests from  <a href="https://github.com/web-platform-tests/wpt/commit/57308167a6128fbf7e2c80d2f5e67321f8c48eff">https://github.com/web-platform-tests/wpt/commit/57308167a6128fbf7e2c80d2f5e67321f8c48eff</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-002.html: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::getAccentColor):
(WebCore::RenderThemeAdwaita::paintMenuList):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustButtonLikeControlStyle const):
(WebCore::RenderThemeIOS::controlTintColor const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::usedAccentColor const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/280631@main">https://commits.webkit.org/280631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b63409e3f7bd5aa94c17afb0cada23978a7e3f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36407 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9554 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60699 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7522 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59207 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44031 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46225 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5293 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34184 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49267 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27086 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30965 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6599 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6527 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52919 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62380 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6971 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53482 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/996 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49322 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53532 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12635 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/839 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32236 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33321 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34406 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33067 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/rendering/ios/RenderThemeIOS.mm:
```diff
@@ -897,7 +897,7 @@ static bool renderThemePaintSwitchTrack(OptionSet<ControlStyle::State>, const Re
 
     if (!style.hasAutoAccentColor()) {
-        auto tintColor = style.usedAccentColor();
+        auto tintColor = style.usedAccentColor(element.renderer()->styleColorOptions());
```
does this need to make sure `renderer` is non null?

nt1m replied:
I would hope that it's non-null here, if there's no renderer, there shouldn't be any need to run this code

pxlcoder replied:
Would use `element.document().styleColorOptions(&style)` instead.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3093,15 +3094,22 @@ Color RenderStyle::colorWithColorFilter(const StyleColor& color) const
     if (hasAutoAccentColor())
         return { };
 
+    Color resolvedAccentColor = colorResolvingCurrentColor(accentColor());
```
```suggestion
    auto resolvedAccentColor = colorResolvingCurrentColor(accentColor());
```

rr-codes replied:
instead of `Color`, you can just `auto`

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -3093,15 +3094,22 @@ Color RenderStyle::colorWithColorFilter(const StyleColor& color) const
+    Color resolvedAccentColor = colorResolvingCurrentColor(accentColor());
+
+    if (!resolvedAccentColor.isOpaque()) {
+        Color computedCanvasColor = RenderTheme::singleton().systemColor(CSSValueCanvas, styleColorOptions);
```
```suggestion
        auto computedCanvasColor = RenderTheme::singleton().systemColor(CSSValueCanvas, styleColorOptions);
```

nt1m reviewed (approved):
Looks good, I would just re-import tests before landing this.

lilyspiniolas committed:
[css-ui-4] Transparent accent colors (275767)
https://bugs.webkit.org/show_bug.cgi?id=275767
rdar://130599744

Reviewed by Tim Nguyen and Abrar Rahman Protyasha.

Transparent accent-colors are now precomposed over the canvas to conform to https://github.com/w3c/csswg-drafts/issues/9852#issuecomment-2166096835. Added transparent-accent-color wpt tests from  https://github.com/web-platform-tests/wpt/commit/57308167a6128fbf7e2c80d2f5e67321f8c48eff

* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/transparent-accent-color-002.html: Added.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::paint):
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
(WebCore::getAccentColor):
(WebCore::RenderThemeAdwaita::paintMenuList):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustButtonLikeControlStyle const):
(WebCore::RenderThemeIOS::controlTintColor const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::usedAccentColor const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: https://commits.webkit.org/280631@main
